File: ts/2025/day03.ts
let pickK = (s: string, n: number): number => {
	let res: string[] = Array(n).fill("0")
	for (let i = n, sta = 0; i > 0; i--) {
		let p = sta, d = "0"
		for (let j = sta; j < s.length - i + 1; j++) {
			const val = s[j]!
			if (val > d) {
				p = j, d = val
			}
		}
		res[n - i] = d
		sta = p + 1
	}
	return Number(res.join(""))

}

let p1 = 0, p2 = 0
for await (const line of console) {
	if (line === "") { continue }
	p1 += pickK(line, 2)
	p2 += pickK(line, 12)
}
console.log("Part 1: %d", p1)
console.log("Part 2: %d", p2)

File: py/2025/day03.py
import sys


def pick(s: str, n: int) -> int:
    res: list[str] = ["0"] * n
    sta: int = 0
    for i in range(n, 0, -1):
        p, d = sta, "0"
        for j in range(sta, len(s) - i + 1):
            if (v := s[j]) > d:
                p, d = j, v
        res[n - i] = d
        sta = p + 1
    return int("".join(res))


p1, p2 = 0, 0
D = sys.stdin.read()
lines = [line for line in D.split("\n") if line]
for line in lines:
    p1 += pick(line, 2)
    p2 += pick(line, 12)

print(f"Part 1: {p1}")
print(f"Part 2: {p2}")
